File: ricecooker.go
package main

import (
	"fmt"
	"strconv"
	"time"
)

// RiceCooker represents a rice cooker.
type RiceCooker struct {
	power bool
}

// checkPower checks the power status of the rice cooker.
func (rc *RiceCooker) checkPower() bool {
	return rc.power
}

// turnOn turns on the rice cooker.
func (rc *RiceCooker) turnOn() {
	rc.power = true
	fmt.Println("Hello!! You are on the rice cooker. The rice cooker is turned on.")
}

// turnOff turns off the rice cooker.
func (rc *RiceCooker) turnOff() {
	rc.power = false
	fmt.Println("The rice cooker is turned off.")
}

// chooseFunction handles user input to perform actions on the rice cooker.
func (rc *RiceCooker) chooseFunction(userInput string) {
	switch userInput {
	case "1":
		rc.turnOn()
		action, err := rc.getValidNumber("What do you want to do? (1 to cook / 2 to boil): ", []int{1, 2})
		if err != nil {
			fmt.Println("Error:", err)
			return
		}
		switch action {
		case 1:
			duration, err := rc.getValidNumber("How many minutes do you want to cook? ", nil)
			if err != nil {
				fmt.Println("Error:", err)
				return
			}
			temperature, err := rc.getValidNumber("At what temperature do you want to cook? ", nil)
			if err != nil {
				fmt.Println("Error:", err)
				return
			}
			rc.cook(duration, temperature)
		case 2:
			rc.boil()
		default:
			fmt.Println("Unrecognized action.")
		}
	case "2":
		rc.turnOff()
	default:
		fmt.Println("Unrecognized command.")
	}
}

// getValidNumber prompts the user for a valid number and validates the input.
func (rc *RiceCooker) getValidNumber(promptMessage string, validChoices []int) (int, error) {
	var userInput string
	for {
		fmt.Print(promptMessage)
		fmt.Scanln(&userInput)
		number, err := strconv.Atoi(userInput)
		if err != nil {
			fmt.Println("Error: Please enter a valid number.")
			continue
		}
		if len(validChoices) > 0 {
			valid := false
			for _, choice := range validChoices {
				if number == choice {
					valid = true
					break
				}
			}
			if !valid {
				fmt.Println("Error: Please enter a valid number among the specified choices.")
				continue
			}
		}
		return number, nil
	}
}

// cook simulates cooking for a specified duration and temperature.
func (rc *RiceCooker) cook(duration, temperature int) {
	fmt.Printf("Cooking in progress for %d minute(s) at %d°C...\n", duration, temperature)
	rc.sleep(duration)
	fmt.Println("Cooking completed!")
}

// boil simulates boiling water.
func (rc *RiceCooker) boil() {
	fmt.Println("Boiling water in progress...")
	boilingTime := 10 // Simulation duration (10 minutes for example)
	currentTemperature := 0

	for i := 1; i <= boilingTime; i++ {
		currentTemperature += 10
		rc.sleep(1)
		fmt.Printf("Water temperature: %d°C - Remaining time: %d minutes\n", currentTemperature, boilingTime-i)
	}

	fmt.Println("Water has reached 100°C. Automatic shutdown.")
}

// sleep pauses execution for the specified number of minutes.
func (rc *RiceCooker) sleep(minutes int) {
	milliseconds := int64(minutes) * 60 * 1000
	startTime := currentTimeMillis()
	for currentTimeMillis() < startTime+milliseconds {
	}
}

// currentTimeMillis returns the current time in milliseconds.
func currentTimeMillis() int64 {
	return time.Now().UnixNano() / int64(time.Millisecond)
}

func main() {
	myRiceCooker := &RiceCooker{}

	fmt.Print("Hello! You are on the rice cooker. Do you want to turn it on? (1 for yes / 2 for no): ")
	userDecision, err := myRiceCooker.getValidNumber("", []int{1, 2})
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	myRiceCooker.chooseFunction(strconv.Itoa(userDecision))
}

File: ricecooker.js
const prompt = require("prompt-sync")();

class RiceCooker {
  constructor() {
    this.power = false;
  }

  checkPower() {
    return this.power;
  }

  turnOn() {
    this.power = true;
    console.log("Hello! You are on the rice cooker. The rice cooker is turned on.");
  }

  turnOff() {
    this.power = false;
    console.log("The rice cooker is turned off.");
  }

  chooseFunction(userInput) {
    switch (userInput) {
      case "1":
        this.turnOn();
        try {
          const action = this.getValidNumber("What do you want to do? (1 to cook / 2 to boil): ", [1, 2]);
          switch (action) {
            case 1:
              const duration = this.getValidNumber("How many minutes do you want to cook? ");
              const temperature = this.getValidNumber("At what temperature do you want to cook? ");
              this.cook(duration, temperature);
              break;

            case 2:
              this.boil();
              break;

            default:
              console.log("Unrecognized action.");
          }
        } catch (error) {
          console.error("Error:", error.message);
        }
        break;

      case "2":
        this.turnOff();
        break;

      default:
        console.log("Unrecognized command.");
    }
  }

  getValidNumber(promptMessage, validChoices = []) {
    let userInput;
    while (true) {
      try {
        userInput = parseInt(prompt(promptMessage), 10);
        if (isNaN(userInput) || (validChoices.length > 0 && !validChoices.includes(userInput))) {
          throw new Error("Please enter a valid number among the specified choices.");
        }
        break;
      } catch (error) {
        console.error("Error:", error.message);
      }
    }
    return userInput;
  }

  cook(duration, temperature) {
    try {
      console.log(`Cooking in progress for ${duration} minute(s) at ${temperature}°C...`);
      this.sleep(duration);
      console.log("Cooking completed!");
    } catch (error) {
      console.error("Error:", error.message);
    }
  }

  boil() {
    try {
      console.log("Boiling water in progress...");
      const boilingTime = 10; // Simulation duration (10 minutes, for example)
      let currentTemperature = 0;

      for (let i = 1; i <= boilingTime; i++) {
        currentTemperature += 10;
        this.sleep(1);
        console.log(`Water temperature: ${currentTemperature}°C - Remaining time: ${boilingTime - i} minutes`);
      }

      console.log("Water has reached 100°C. Automatic shutdown.");
    } catch (error) {
      console.error("Error:", error.message);
    }
  }

  sleep(minutes) {
    const milliseconds = minutes * 60 * 1000;
    const startTime = new Date().getTime();
    while (new Date().getTime() < startTime + milliseconds);
  }
}

// Example usage
const myRiceCooker = new RiceCooker();

try {
  const userDecision = myRiceCooker.getValidNumber(
    "Hello! You are on the rice cooker. Do you want to turn it on? (1 for yes / 2 for no): ",
    [1, 2]
  );
  myRiceCooker.chooseFunction(userDecision.toString());
} catch (error) {
  console.error("Error:", error.message);
}
